mmdet/models/roi_heads/sparse_roi_head.py

```diff
@@ -0,0 +1,313 @@
+        """Forward function in training stage.
+
+        Args:
+            x (list[Tensor]): list of multi-level img features.
+            proposals (Tensor): Decoded proposal bboxes, has shape
+                (batch_size, num_proposals, 4)
+            proposal_features (Tensor): Expanded proposal
+                features, has shape
+                (batch_size, num_proposals, proposal_feature_channel)
+            img_metas (list[dict]): list of image info dict where
+                each dict has: 'img_shape', 'scale_factor', 'flip',
+                 and may also contain 'filename', 'ori_shape',
+                 'pad_shape', and 'img_norm_cfg'. For details on the
+                 values of these keys see
+                `mmdet/datasets/pipelines/formatting.py:Collect`.
+            gt_bboxes (list[Tensor]): Ground truth bboxes for each image with
+                shape (num_gts, 4) in [tl_x, tl_y, br_x, br_y] format.
+            gt_labels (list[Tensor]): class indices corresponding to each box
+            gt_bboxes_ignore (None | list[Tensor]): specify which bounding
+                boxes can be ignored when computing the loss.
+            imgs_whwh (Tensor): Tensor with shape (batch_size, 4),
+                    the dimension means
+                    [img_width,img_height, img_width, img_height].
+            gt_masks (None | Tensor) : true segmentation masks for each box
+                used if the architecture supports a segmentation task.
+
+        Returns:
+            dict[str, Tensor]: a dictionary of loss components of all stage.
+        """
+
+        num_imgs = len(img_metas)
+        num_poposals = proposal_boxes.size(1)
+        imgs_whwh = imgs_whwh.repeat(1, num_poposals, 1)
+        all_stage_bbox_results = []
+        proposal_list = [proposal_boxes[i] for i in range(len(proposal_boxes))]
+        object_feats = proposal_features
+        all_stage_loss = {}
+        for stage in range(self.num_stages):
+            rois = bbox2roi(proposal_list)
+            bbox_results = self._bbox_forward(stage, x, rois, object_feats,
+                                              img_metas)
+            all_stage_bbox_results.append(bbox_results)
+            if gt_bboxes_ignore is None:
+                # TODO support ignore
+                gt_bboxes_ignore = [None for _ in range(num_imgs)]
+            sampling_results = []
+            cls_pred_list = bbox_results['detach_cls_score_list']
+            proposal_list = bbox_results['detach_proposal_list']
+            for i in range(num_imgs):
+                normolize_bbox_ccwh = bbox_xyxy_to_cxcywh(proposal_list[i] /
+                                                          imgs_whwh[i])
+                assign_result = self.bbox_assigner[stage].assign(
+                    normolize_bbox_ccwh, cls_pred_list[i], gt_bboxes[i],
+                    gt_labels[i], img_metas[i])
+                assert isinstance(self.bbox_sampler[stage], PseudoSampler), \
```

Comment:
Should we assert the config's type during initialization rather than at running time?